File: 1834-minimum-number-of-people-to-teach/minimum-number-of-people-to-teach.cpp
class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        unordered_set<int>sadUsers;

        for (auto fri:friendships){
            int u=fri[0]-1;
            int v=fri[1]-1;
            unordered_set<int>lang(languages[u].begin(),languages[u].end());
            bool flag=false;
            for (auto l:languages[v]){
                if (lang.count(l)){
                    flag=true;
                    break;
                }
            }
            if (!flag) {sadUsers.insert(u); sadUsers.insert(v);}

        }

        int maxKnown=0;

        vector<int>freq(n+1,0);

        for (auto user:sadUsers){
            for (auto l:languages[user]){
                freq[l]++;
                maxKnown=max(maxKnown,freq[l]);
            }
        }
        return sadUsers.size()-maxKnown;
    }
};

File: 1834-minimum-number-of-people-to-teach/minimum-number-of-people-to-teach.java
class Solution {
    public int minimumTeachings(int n, int[][] languages, int[][] friendships) {
        Set<Integer>sadUsers=new HashSet<>();

        for (int [] friend:friendships){
            int u=friend[0]-1;
            int v=friend[1]-1;

            Set<Integer>lang=new HashSet<>();

            for (int l:languages[u]){
                lang.add(l);
            }

            boolean canTalk=false;

            for (int l:languages[v]){
                if (lang.contains(l)){
                    canTalk=true;
                    break;
                }
            }
            if (canTalk==false){
                sadUsers.add(u);
                sadUsers.add(v);
            }
        }

        int maxKnown=0;

        int[] freq=new int[n+1];

        for(int user:sadUsers){
            for (int l:languages[user]){
                freq[l]++;
                maxKnown=Math.max(maxKnown,freq[l]);
            }
        }
        return sadUsers.size()-maxKnown;
    }
}